### DISCUSSION TITLE
Clarify the language in entry.h about CLAP_PATH

### DISCUSSION THREAD

**Discussion by baconpaul** [free-audio/clap#102] [General]
I suggest two clarifications to the language in `entry.h` following the conversation here.

https://www.kvraudio.com/forum/viewtopic.php?t=583188&sid=2b626a18c4f414326a0a845c078b0091

The current 1.0.2 language is

```
// Additionally, extra path may be specified in CLAP_PATH environment variable.
// CLAP_PATH is formated in the same way as the OS' binary search path (PATH on UNIX, Path on Windows).
```

this language could be improved in several ways

1: The concept of 'extra path as os path' is clear to people who know how an OS path works, but not clear to others.
2: It is unclear, and therefore ambiguous, whether a host should search a directory recursively or whether it should simply look in a member of the `CLAP_PATH`.  I am proposing that we imply a recursive search down each member of the PATH, but we can adjust the language
3: It is unclear what responsibility a plugin installer has if it chooses to install in a non-standard location. I suggest that we encourage such installers to update CLAP_PATH if the install location is not already in CLAP_PATH.

Therefore I suggest this language become

```
In addition to the OS-specific default locations above, a CLAP_HOST must query the environment
for a CLAP_PATH variable, which is a list of directories formatted in the same manner as the host OS
binary search path (PATH on Unix, separated by `:` and Path on Windows, separated by ';', as of this writing).

Each directory in the CLAP_PATH should be recursively searched for files and/or bundles as appropriate in your OS
ending with the extension `.clap`.

Moreover, we strongly suggest that a plugin installer install in a location which is either in the standard locations
or is in the CLAP_PATH, and if it offers alternate locations to users, guarantee that at the end of the install,
the location chosen is either a standard location or modifies the CLAP_PATH to indicate the install location to hosts.
```

**Comment by tim-janik** [comment]
As described in my latest [KVR Post](https://www.kvraudio.com/forum/viewtopic.php?p=8453410#p8453410), recursive scanning is necessary (though the spec currently fails to outline this), because some plugins install into subdirectories:
```sh
~$ ls .clap/*
.clap/u-he:
ACE.64.clap
Diva.64.clap
Hive.64.clap
MFM2.64.clap
```
Useful clarification would include whether "hidden" subdirectories can be omitted from scanning, e.g. `.samples/` or `.icons/` so plugins may ship and install resources in subdirectories without causing useless subtree scans by the hosts.

**Reply by baconpaul** [reply]
Agree we need clarity on hidden dirs

**Reply by abique** [reply]
The host would only scan the files matching `*.clap`.
What's the benefit in preventing to scan hidden directories?

**Reply by baconpaul** [reply]
Scan time mostly. I believe reaper does this for instance

**Reply by abique** [reply]
I'd be curious to know how much time you'd save on a SSD. I expect that it would justify the complexity introduced in the spec.
On Linux I often do massive searches or even `grep -r` and that's fast on my 7 years old computer. Maybe Windows is a different game but still, I have doubt it would help.

**Reply by tim-janik** [reply]
> The host would only scan the files matching `*.clap`. What's the benefit in preventing to scan hidden directories?

For instance the u-he/Hive plugin consists of over 6000 files in 236 subdirectories. Of this, only `Hive.64.so` is worth scanning by the host. Now the u-he installation is fairly clever in putting it all into ~/.u-he/Hive/ and then installing a single symlink from ~/.clap/u-he/Hive.64.clap into ~/.u-he/Hive/. If it was putting the entire Hive/ directory under ~/.clap/ that would slow down the scanning for `*.clap` files by orders of magnitudes.

So unless the spec demands that clap plugins should only be installed under clap/ as a symlink to binaries (which might not work as well under Windows or with some hosts that tend to ignore symlinks alltogether), it'd be wise to recommend a combination similar to e.g.:
1) Hidden directories should be omitted during recursive scans (under Unix, hidden directories start with a dot);
2) Plugin installations should move extra resource files into hidden subdirectories, which are omitted during recursive scans by CLAP hosts.

**Reply by abique** [reply]
```
22-06-17 09:43:36 abique@knot:~/d/c/b/ninja-system git:main> time find /home/abique/.u-he/ | wc -l
37949

real	0m0.030s
user	0m0.015s
sys	0m0.019s
22-06-17 09:43:39 abique@knot:~/d/c/b/ninja-system git:main> 
```

**Reply by abique** [reply]
Really I don't see the need to make the spec more complicated for that.

**Reply by abique** [reply]
I tried something even bigger:
```
22-06-17 09:45:49 abique@knot:~/d/c/b/ninja-system git:main> time find /usr/ 2>/dev/null | wc -l
506538

real	0m1.318s
user	0m0.221s
sys	0m0.500s
22-06-17 09:46:39 abique@knot:~/d/c/b/ninja-system git:main> 
```
So it took 1.3s to list my entire /usr consisting of 500K files on my 7 years old computer (which can't even install windows 11).

**Reply by baconpaul** [reply]
OK!

**Comment by abique** [comment]
```
Moreover, we strongly suggest that a plugin installer install in a location which is either in the standard locations
or is in the CLAP_PATH, and if it offers alternate locations to users, guarantee that at the end of the install,
the location chosen is either a standard location or modifies the CLAP_PATH to indicate the install location to hosts.
```
I'm against modifying `CLAP_PATH` for the user. I'd be horrified if a plugin or anything changes my environment variable. I don't trust and I would never let anyone touch my bashrc or my configuration files.

I think the right approach is:
- offer to install in default location, system wide or in the user's home
- otherwise let the user specify its directory and be responsible for setting up his DAW correctly, may that be using `CLAP_PATH` or just the DAW settings.

What do you think?

**Reply by abique** [reply]
```
In addition to the OS-specific default locations above, a CLAP_HOST must query the environment
for a CLAP_PATH variable, which is a list of directories formatted in the same manner as the host OS
binary search path (PATH on Unix, separated by `:` and Path on Windows, separated by ';', as of this writing).

Each directory in the CLAP_PATH should be recursively searched for files and/or bundles as appropriate in your OS
ending with the extension `.clap`.
```
@baconpaul this part is very good :-)

**Reply by baconpaul** [reply]
Yeah right so over on KVR I made the point that: A correct installer won't leave your system in an invalid state.

And installing a clap in a directory which isn't in the CLAP_PATH or the specified areas is an invalid state.

Which is why I throught: add an option to update clap_path

we could have the spec be silent in this regard and instead add a sentence saying 

```
A .clap which is installed in a  directory which is neither the enumerated fix directories or a descendant of a member of 
the install-time  CLAP_PATH is installed in an invalid location. We encourage plugin installers to not allow such 
configurations through any of a variety of mechanisms, including reading CLAP_PATH, helping users modify CLAP_PATH if
appropriate, or installing to fixed locations.
```

I think we should also have a hidden directory carve out for the recursion. 

```
Well-formed hosts may opt to not recurse into hidden directories (directories starting with a '.' on Linux, or directories with hidden
OS level flags on macOS and windows).
```

**Reply by baconpaul** [reply]
also would love feedback from @Bremmers on this

**Reply by abique** [reply]
I think that if the plugin allows the user to install in non-standard directory then it is up to the user and plugin installer to figure out the rest.

To me non standard installations falls back into the `I KNOW WHAT I'M DOING` situation. So the user is then responsible for his system and daw configuration.

**Reply by Bremmers** [reply]
> also would love feedback from @Bremmers on this

Frankly, I think CLAP_PATH is a bad idea that will cause problems. Let's remove it while we can.

In any case, plugins shouldn't set it themselves. Perhaps hosts shouldn't set it either, so only tech-geek users can use this. Plugin installers should just use the default location if CLAP_PATH isn't set.

**Reply by abique** [reply]
`CLAP_PATH` is still useful.
But I agree that a basic user should not be concerned at all with it, ideally he should not even know its existence.
It is so easy to add an extra location in your DAW, while adding an extra environment variable is difficult.

**Reply by abique** [reply]
@baconpaul I've used your two first paragraphs, but I think we don't need to say anything regarding installers. https://github.com/free-audio/clap/pull/106

**Reply by robbert-vdh** [reply]
Yes `CLAP_PATH` should definitely not be modified automatically, It's a feature that exists for power users, and it should not be too visible for 'normal' users. But for those power users it still needs to exist, and if it's not defined by the spec then everyone will just come up with its own solution. Similar things have already happened with other plugin specifications, so having it be standardized like this is a good idea IMO.

**Reply by baconpaul** [reply]
I agree.

**Comment by jpcima** [comment]
If I may ask regarding bundles:

a fair bit ago, I asked on KVR about whether CLAP can support bundles in VST3's fashion;
this was answered with yes, but there seems no explicit info about this anywhere.

Could one install a file structure like this in a CLAP folder and expect it to work on windows and linux OS?
```
my.clap/Contents/Binary/my.clap
my.clap/Contents/Resources/sample.wav
```

**Reply by robbert-vdh** [reply]
Although that's not part of the standard, there's no reason why that wouldn't work since they're just directories. Although on Linux and Windows you may want to skip the ceremony and just do:

```
supercooldevname/my/my.clap
supercooldevname/my/sample.wav
```

**Reply by jpcima** [reply]
>Although on Linux and Windows you may want to skip the ceremony and just do

Yes, except in case of using vstgui. This library forces to use bundle-like structure.
It goes from the module handle's location, up 2 folders, and into "Resources".

**Reply by baconpaul** [reply]
If you use the resource mechanism that is. If you don’t then vstgui works just fine, as you know of course 

bundles work so poorly on win and lin. But if we want to change the standard so a .clap can be a file or a directory on win lin I would not object.

**Reply by baconpaul** [reply]
With the associated description of win lin bundle layout, of course. Macos we would use standard bundles of course

**Reply by robbert-vdh** [reply]
I don't think the standard would need to be changed, right? That's also the reason why the 'new' style bundles on Windows still use the `.vst3` file extension. Old hosts don't need to be updated to handle the new bundles as they'll just see a `.vst3` file. So if you put a `.clap` file inside of a bundle like that on Windows then it will just look like any other CLAP plugin to the host.

**Reply by baconpaul** [reply]
Right. But you can’t call your top of bundle .clap
As we have just found out with a surge installer bug 
So I think we should explain that. And maybe suggest a suffix like .clapbundle 
Or maybe that’s a bad idea :)

**Reply by robbert-vdh** [reply]
I don't see why you wouldn't be able to do that? `foo.clap` is a perfectly valid directory name. Not that I'd encourage people to do that, of course. :wink:

**Reply by baconpaul** [reply]
because people stop recursing at anything.clap

**Reply by baconpaul** [reply]
so we would have to say in the spec that "on windows and linux a clap file must be a regular file; and may exist in a directory which itself has the name .clap; on macOS a .clap file is a directory bundle"

and "people" is "bws 43b7 on windows" https://github.com/surge-synthesizer/surge/issues/6291#issuecomment-1159486454

**Reply by baconpaul** [reply]
we really need a reference implantation of a travers-and-scan. I started one in clap-info last night. i believe it would stop on a .clap directory on win/lin though and i should fix that.

**Reply by robbert-vdh** [reply]
Then the scanning logic in those hosts are broken. And then we'd be seeing the same issue with VST3 on Windows. When walking a directory to scan for plugins you'd recurse through all directories and symlinks to directories, and then find all non-directory files in those directories with a `.clap` extension. I don't think anyone would specifically add a filter to stop recursing after a directory that ends in `.clap`.

**Reply by baconpaul** [reply]
Or in other words, when scanning for .clap files a correct host does this

```cpp
            for (auto const &dir_entry : std::filesystem::recursive_directory_iterator(p))
            {
#if MAC
                if (dir_entry.path().extension().u8string() == ".clap" &&
                    std::filesystem::is_directory(dir_entry.path()))
                {
                    claps.emplace_back(dir_entry.path());
                }
#else
                if (dir_entry.path().extension().u8string() == ".clap" &&
                    std::filesystem::is_regular_file(dir_entry.path()))
                {
                    claps.emplace_back(dir_entry.path());
                }
#endif
      }
```

**Reply by robbert-vdh** [reply]
The `std::filesystem::is_regular_file` line should be `(std::filesystem::is_regular_file(dir_entry.path())) || std::filesystem::is_symlink(dir_entry.path())))`. Or for most practical purposes, a `!std::filesystem::is_directory` is probably also good enough.

**Reply by jpcima** [reply]
Can it also be clarified what is the `plugin_path` provided by `init`?
does the point to the executable or bundle, on the Mac platform?

**Reply by baconpaul** [reply]
Good point JP. The path points to the renamed dll on lin win and the bundle on mac indeed
